### DISCUSSION TITLE
Repeat, Remove & Change schedule

### DISCUSSION THREAD

**Discussion by neandr** [jensrossbach/node-red-contrib-chronos#111] [Q&A]
When I first read the documentation, I understood that a "scheduler" setup is automatically repeated every day. Correct?
How can a complete "scheduler" be disabled?
Is it possible to remove (not just disable) specific defined "Schedule" tasks? Or would it be necessary to remove the entire "Scheduler" and restore a new "Scheduler" set with the desired changes?
Can a specific defined Schedule task be addressed only by its position in the "Schedule" or is there another method?

Overall the Chronos contribution looks great!

**Comment by jensrossbach** [comment]
Hi @neandr!

> When I first read the documentation, I understood that a "scheduler" setup is automatically repeated every day. Correct?

This is correct for all schedule types except the cron table schedule type. The latter can have trigger intervals in the range of years up to the range of seconds, see also [here](https://github.com/jensrossbach/node-red-contrib-chronos/wiki/Scheduler-Node#cron-table).

> How can a complete "scheduler" be disabled?

You can disable the whole scheduler node right from the beginning (i.e., flow deploy or Node-RED start) by checking the option _Start with disabled schedule_ in the configuration of a node or you can disable it at runtime by sending a message with a `msg.payload` set to `false`, see also [here](https://github.com/jensrossbach/node-red-contrib-chronos/wiki/Scheduler-Node#enabledisable-schedule).

> Is it possible to remove (not just disable) specific defined "Schedule" tasks? Or would it be necessary to remove the entire "Scheduler" and restore a new "Scheduler" set with the desired changes?

You cannot remove a schedule, you can only disable or override a schedule, see also [here](https://github.com/jensrossbach/node-red-contrib-chronos/wiki/Scheduler-Node#enabledisable-schedule) and [here](https://github.com/jensrossbach/node-red-contrib-chronos/wiki/Scheduler-Node#control-scheduler-node).

> Can a specific defined Schedule task be addressed only by its position in the "Schedule" or is there another method?

Schedules can only be addressed by their 0-based position from input messages.

**Reply by neandr** [reply]
Great, gives some clarification for using it with my (planned) project  ;)

**Comment by neandr** [comment]
Again for better understanding:
If you don't want the schedule to run automatically the next day, just pass an empty array or string with the msg.payload.
On the other hand, a new modified schedule array would overwrite the previously set elements and it is not sure that all elements have the "new" desired setting.
Therefore, it is best to send an empty array or string as msg.payload to clear all previously active elements and only then send the new elements in a second msg.payload.

**Reply by jensrossbach** [reply]
I'm not sure if I understand the use case. If you simply want a schedule not to run the next day, you can disable it (sending `msg.payload` as array with `false` at the index of your schedule or complete `msg.payload` as boolean `false` to disable all schedules). Sending an empty array or string does not make sense at all as it is not a valid input and would do nothing.
If you want to override the schedules on the other hand, you have to send an array as `msg.payload` with each array element being an object containing the schedule data. If you want to override only some of the schedules, you can set array elements to `null` for these schedules that you do not want to override.

Maybe you can explain your exact use case a bit more in detail because I'm not sure if above information is what you need. What is the purpose for overriding the schedules in your use case?

**Reply by neandr** [reply]
For time-controlled switching and control of lighting fixtures in the house I use a Python-based program on a RPI. The devices with Tasmota firmware are controlled via HTTP.
The main features of the current solution are:
-- creation of daily schedules that can take into account sunrise/sunset,
-- all time settings can vary randomly
-- different daily schedules for weekdays (Mon-Fri/Sat/Sun)
-- suspend/reactivate the day schedule with a UI button
-- and some other features

This should be replaced by a NodeRED project. The Cronos project is very close to the requirements, only unclear at the moment is how to change the daily schedule. A "changed" daily schedule can contain not only different times, but also different devices.

**Reply by jensrossbach** [reply]
What about using a combination of scheduler node and time filter nodes for the daily schedule?

For example you can configure three schedules with dedicated output ports, one for Mon-Fri, one for Sat and one for Sun. Then you connect a time filter node to each of the output ports of the scheduler node. In the fist time filter node, you add a condition for Mon, Tue, Wed, Thu and Fri, in the second one you add one for Sat and in the last one, you add one for Sun.
